File: sdk/spring-boot-starter-test/src/main/scala/kalix/spring/boot/KalixConfigurationTest.scala
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.spring.boot

import com.typesafe.config.Config
import kalix.javasdk.JsonSupport
import kalix.javasdk.impl.JsonMessageCodec
import kalix.javasdk.testkit.KalixTestKit
import kalix.spring.impl.KalixSpringApplication
import kalix.spring.impl.WebClientProviderHolder
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.AutoConfiguration
import org.springframework.context.ApplicationContext
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Lazy
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.http.codec.json.Jackson2JsonEncoder
import org.springframework.web.reactive.function.client.WebClient

/**
 * Auto-configuration for Kalix TestKit.
 *
 * When in the classpath, it disables the default KalixConfiguration that would normally start the Kalix Runtime.
 *
 * This configuration is also marked as `@Lazy` to avoid starting the Kalix Runtime automatically. The KalixTestKit, and
 * by consequence the Kalix Runtime, will only start if a test requires it.
 *
 * For example, a test extending [[kalix.spring.testkit.KalixIntegrationTestKitSupport]] will automatically trigger the
 * initialization of the testkit, but a regular test will not. This allows for a more flexible setup in case we want
 * bootstrap the services maually.
 */
@Lazy
@AutoConfiguration
class KalixConfigurationTest(applicationContext: ApplicationContext) extends KalixConfiguration(applicationContext) {

  private val logger = LoggerFactory.getLogger(getClass)

  /** WebClient pointing to the proxy. */
  @Bean
  def createWebClient(kalixTestKit: KalixTestKit): WebClient = WebClient.builder
    .baseUrl("http://localhost:" + kalixTestKit.getPort)
    .defaultHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
    .codecs(configurer => {
      configurer.defaultCodecs.jackson2JsonEncoder(
        new Jackson2JsonEncoder(JsonSupport.getObjectMapper, MediaType.APPLICATION_JSON))
    })
    .build

  @Bean
  def testkitSettings(@Autowired(required = false) settings: KalixTestKit.Settings): Option[KalixTestKit.Settings] =
    Option(settings)

  @Bean
  def kalixTestKit(
      kalixSpringApplication: KalixSpringApplication,
      config: Config,
      settingsOpt: Option[KalixTestKit.Settings]): KalixTestKit = {

    val settings = settingsOpt.getOrElse(KalixTestKit.Settings.DEFAULT)
    val kalixTestKit = new KalixTestKit(kalixSpringApplication.kalix, new JsonMessageCodec, settings)

    logger.info(s"Starting Kalix TestKit with: $settings")
    kalixTestKit.start(config)

    val holder = WebClientProviderHolder.get(kalixTestKit.getRunner.system);
    //when ComponentClient is used in integration test, we must initiate webclient before the first request
    kalixSpringApplication.kalixClient.setWebClient(holder.webClientProvider.localWebClient)

    logger.info(s"Kalix Runtime running on port: ${kalixTestKit.getPort}")
    kalixTestKit
  }

}

File: sdk/java-sdk-spring/src/it/java/kalix/spring/KalixConfigurationTest.java
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.spring;

import kalix.javasdk.JsonSupport;
import kalix.javasdk.client.ComponentClient;
import kalix.javasdk.impl.client.ComponentClientImpl;
import kalix.javasdk.impl.JsonMessageCodec;
import kalix.javasdk.testkit.KalixTestKit;
import kalix.spring.boot.KalixConfiguration;
import kalix.spring.impl.KalixSpringApplication;
import kalix.spring.impl.WebClientProviderHolder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.context.ApplicationContext;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Import;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.codec.json.Jackson2JsonEncoder;
import org.springframework.web.reactive.function.client.WebClient;

@Import(KalixConfiguration.class)
@TestConfiguration
public class KalixConfigurationTest {

  private Logger logger = LoggerFactory.getLogger(getClass());

  @Autowired
  private ApplicationContext applicationContext;
  @Autowired
  private KalixConfiguration kalixConfiguration;

  @Bean
  public KalixSpringApplication kalixSpringApplication() {
    // for this internal integration tests, we need to explicitly pass the Main class we want
    // because otherwise it will detect sbt.ForkMain instead
    return new KalixSpringApplication(applicationContext, kalixConfiguration.config());
  }

  /**
   * WebClient pointing to the proxy.
   */
  @Bean
  public WebClient createWebClient(KalixTestKit kalixTestKit) {
    return WebClient.builder()
      .baseUrl("http://localhost:" + kalixTestKit.getPort())
      .defaultHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
      .codecs(configurer -> {
        configurer.defaultCodecs().jackson2JsonEncoder(
          new Jackson2JsonEncoder(JsonSupport.getObjectMapper(), MediaType.APPLICATION_JSON));
      })
      .build();
  }

  @Bean
  public KalixTestKit kalixTestKit(KalixSpringApplication kalixSpringApplication, KalixTestKit.Settings settings) {
    logger.info("Starting Kalix TestKit...");
    KalixTestKit kalixTestKit = new KalixTestKit(kalixSpringApplication.kalix(), new JsonMessageCodec(), settings);
    kalixTestKit.start(kalixConfiguration.config());
    logger.info("Kalix Runtime running on port: " + kalixTestKit.getPort());
    //when ComponentClient is used in integration test, we must initiate webclient before the first request
    WebClientProviderHolder holder = WebClientProviderHolder.get(kalixTestKit.getRunner().system());
    kalixSpringApplication.kalixClient().setWebClient(holder.webClientProvider().localWebClient());
    return kalixTestKit;
  }

  @Bean
  public ComponentClient componentClient(KalixSpringApplication kalixSpringApplication) {
    return new ComponentClientImpl(kalixSpringApplication.kalixClient());
  }
}
